rclcpp/src/rclcpp/parameter_service.cpp

```diff
@@ -73,12 +77,19 @@ ParameterService::ParameterService(
       const std::shared_ptr<rcl_interfaces::srv::SetParameters::Request> request,
       std::shared_ptr<rcl_interfaces::srv::SetParameters::Response> response)
     {
-      std::vector<rclcpp::Parameter> pvariants;
+      // Set parameters one-by-one, since there's no way to return a partial result if
+      // set_parameters() fails.
+      auto result = rcl_interfaces::msg::SetParametersResult();
       for (auto & p : request->parameters) {
-        pvariants.push_back(rclcpp::Parameter::from_parameter_msg(p));
+        try {
+          result = node_params->set_parameters({rclcpp::Parameter::from_parameter_msg(p)})[0];
```

Comment:
Since you are calling them one by one, I'd consider calling `set_parameters_atomically` (if it's more efficient to do so).

Reply:
Makes sense. Updated 6b9e9de